embassy-stm32/src/pka/mod.rs

```diff
@@ -0,0 +1,485 @@
+    PointOnEllipticCurveFpCheck = 0b101000,
+    /// RSA CRT Exponentiation
+    RsaCrtExponentiation = 0b000111,
+    /// Modular Inversion
+    ModularInversion = 0b001000,
+    /// Arithmetic Addition
+    ArithmeticAddition = 0b001001,
+    /// Arithmetic Subtraction
+    ArithmeticSubtraction = 0b001010,
+    /// Arithmetic Multiplication
+    ArithmeticMultiplication = 0b001011,
+    /// Arithmetic Comparison
+    ArithmeticComparison = 0b001100,
+    /// Modular Reduction
+    ModularReduction = 0b001101,
+    /// Modular Addition
+    ModularAddition = 0b001110,
+    /// Modular Subtraction
+    ModularSubtraction = 0b001111,
+    /// Montgomery Multiplication
+    MontgomeryMultiplication = 0b010000,
+}
+
+/// PKA SealedInstance trait.
+///
+/// Allows access to PAC exclusively for `PKA` module.
+trait SealedInstance {
+    fn regs() -> pac::pka::Pka;
+}
+
+/// PKA interrupt
+pub struct InterruptHandler<T: Instance> {
+    __phantom: PhantomData<T>,
+}
+
+impl<T: Instance> interrupt::typelevel::Handler<T::Interrupt> for InterruptHandler<T> {
+    unsafe fn on_interrupt() {
+        let procendf = T::regs().sr().read().procendf();
+        let addr_err = T::regs().sr().read().addrerrf();
+        let ram_err = T::regs().sr().read().ramerrf();
+
+        if procendf {
+            T::regs().cr().modify(|w| w.set_procendie(false));
+        }
+
+        if addr_err {
+            T::regs().cr().modify(|w| w.set_addrerrie(false));
+        }
+
+        if ram_err {
+            T::regs().cr().modify(|w| w.set_ramerrie(false));
+        }
+
+        PKA_WAKER.wake();
+    }
+}
+
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+/// RSA structure for PKA accelerator
+pub struct Rsa<'c, 'd, T: Instance, const RSA_KEY_SIZE: usize>
+where
+    [(); RSA_KEY_SIZE / 4]:,
+{
+    /// Modulus
+    pub modulus: [u8; RSA_KEY_SIZE],
+    /// R^2 mod N
+    r2_mod_n: [u8; RSA_KEY_SIZE],
+    /// public exponent
+    pub public_exponent: [u8; RSA_KEY_SIZE],
+    /// private exponent
+    /// FIXME: In some edge cases the private exponent can be larger than the modulus
+    pub private_exponent: [u8; RSA_KEY_SIZE],
+    /// PKA accelerator
+    pka: &'c mut Pka<'d, T, RSA_KEY_SIZE>,
+}
+
+impl<'c, 'd, T: Instance, const RSA_KEY_SIZE: usize> Rsa<'c, 'd, T, RSA_KEY_SIZE>
+where
+    [(); RSA_KEY_SIZE / 4]:,
+{
+    /// Create a new RSA structure
+    pub fn new(pka: &'c mut Pka<'d, T, RSA_KEY_SIZE>) -> Self {
+        pka.enable();
+        Self {
+            modulus: [0; RSA_KEY_SIZE],
+            r2_mod_n: [0; RSA_KEY_SIZE],
+            public_exponent: [0; RSA_KEY_SIZE],
+            private_exponent: [0; RSA_KEY_SIZE],
+            pka,
+        }
+    }
+
+    /// Reorder data for the new word order
+    fn word_reorder<const N: usize>(input: &[u8; N]) -> [u8; N] {
+        // Ensure the input length is a multiple of 4 by padding with 0s if necessary
+        assert!(N % 4 == 0, "Input length must be a multiple of 4");
```

Comment:
I think in general, whenever possible, we prefer to have an `Error` type and choose to return that without panicking, giving the option for the user to handle it in whichever way is appropriate.

Reply:
This applies to many parts of the code in these drivers, so I will not repeat the comments everywhere.

Reply:
I understand. I will find parts of the code that might panic (like `unwrap`) and convert it to `Errror` return type.

Reply:
It does not